tests/integration/pilot/tunneling/tunneling_test.go

```diff
@@ -0,0 +1,421 @@
+//go:build integ
+// +build integ
+
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tunneling
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/framework/resource"
+	kubetest "istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/tests/integration/pilot/common"
+)
+
+type tunnelingTestCase struct {
+	// name must be unique, because it's used in the requested URL and then it's searched for in the access.log file,
+	// so a duplicated name would make test false positive
+	name string
+	// protocols specifies what types of requests to test; it can contain "http" or "https" values
+	protocols []string
+	// istioResourcesToApply is a list of files with Istio resources required to apply to test a particular case
+	istioResourcesToApply []string
+}
+
+func testCases() []tunnelingTestCase {
+	return []tunnelingTestCase{
+		{
+			"gateway/tcp",
+			[]string{"http", "https"},
+			[]string{
+				"gateway/tcp/virtual-service.yaml",
+				"gateway/tcp/gateway.yaml",
+			},
+		},
+		{
+			"gateway/tls/istio-mutual",
+			[]string{
+				"http",
+				// Temporarily disabled, because of PR #38499 (commit 3643f2c172b91927ac4204f8dd16ba4b00a59832).
+				// sidecar logs:
+				// 2022-05-11T09:32:03.801667Z     warning envoy filter    Incorrect istio-peer-exchange ALPN magic. Peer missing TCP MetadataExchange filter.
+				// [2022-05-11T09:32:02.736Z] "- - -" 0 - - - "-" 517 309 2 - "-" "-" "-" "-" "10.244.0.34:8443" outbound|443|external-app-mtls-gateway|istio-egressgateway.istio-system.svc.cluster.local 10.244.0.33:48312 10.244.0.29:443 10.244.0.33:46102 external-app.testdomain -
+				// "https",
```

Comment:
Tunneling HTTPS request through ISTIO_MUTUAL gateway stopped work, because of #38499. I'm not well familiar with istio-peer-exchange ALPN, so could you advice me how to fix this problem? @vikaschoudhary16 (as you are the author of this change), @hzxuzhonghu (as you approved this change)

Reply:
As we discussed on Slack, #38495 will solve this problem.